artiq/coredevice/phaser.py

```diff
@@ -468,6 +474,32 @@ def get_crc_err(self):
         """
         return self.read8(PHASER_ADDR_CRC_ERR)
 
+    @kernel
+    def measure_frame_timestamp(self):
+        """Perform a register read and record the exact frame timing in `self.frame_tstamp`.
+
+        Deterministic timing requires updates to be schedule at multiples of `self.t_frame` later.
+        See `get_next_frame_timestamp()`.
+        """
+        rtio_output((self.channel_base << 8) | (PHASER_ADDR_BOARD_ID & 0x7f), 0)  # can read any register
+        delay_mu(int64(self.t_frame))
+        self.frame_tstamp = rtio_input_timestamp(rtio_get_counter() + 0xffffff, self.channel_base)
+        delay(10*ms)
+
+    @kernel
+    def get_next_frame_timestamp(self, after_timestamp_mu = int64(-1)):
+        """Return the RTIO timestamp of the next frame after `after_timestamp_mu`.
```

Comment:
You need to define "timestamp of the next frame". This gives only a future timestamp at a deterministic offset w.r.t. the current frame. Also, I'm unsure what the use case for the `after_timestamp_mu` is.

Reply:
The intent was to allow scheduling aligned events in the future. But this is probably not really needed. Removed `after_timestamp_mu` in 075cb26. Is the new docstring more accurate?

Reply:
When scheduling in the future, you would likely do `delay()` anyway which then gets picked up by `now_mu()`.